librosa/core/constantq.py

```diff
@@ -728,10 +730,257 @@ def icqt(C, sr=22050, hop_length=512, fmin=None, bins_per_octave=12,
     return y
 
 
+@cache(level=20)
+def vqt(y, sr=22050, hop_length=512, fmin=None, n_bins=84, gamma=None,
+        bins_per_octave=12, tuning=0.0, filter_scale=1,
+        norm=1, sparsity=0.01, window='hann',
+        scale=True, pad_mode='reflect', res_type=None):
+    '''Compute the variable-Q transform of an audio signal.
+
+    This implementation is based on the recursive sub-sampling method
+    described by [1]_.
+
+    .. [1] Schörkhuber, Christian, Anssi Klapuri, Nicki Holighaus, and Monika Dörfler.
+        "A Matlab toolbox for efficient perfect reconstruction time-frequency transforms with log-frequency resolution."
+        In Audio Engineering Society Conference: 53rd International Conference: Semantic Audio.
+        Audio Engineering Society, 2014.
+
+    Parameters
+    ----------
+    y : np.ndarray [shape=(n,)]
+        audio time series
+
+    sr : number > 0 [scalar]
+        sampling rate of `y`
+
+    hop_length : int > 0 [scalar]
+        number of samples between successive CQT columns.
+
+    fmin : float > 0 [scalar]
+        Minimum frequency. Defaults to C1 ~= 32.70 Hz
+
+    n_bins : int > 0 [scalar]
+        Number of frequency bins, starting at `fmin`
+
+    gamma : number > 0 [scalar]
+        Bandwidth offset for determining filter lengths.
+        `gamma=0` produces the constant-Q transform.
+
+    bins_per_octave : int > 0 [scalar]
+        Number of bins per octave
+
+    tuning : None or float
+        Tuning offset in fractions of a bin.
+
+        If `None`, tuning will be automatically estimated from the signal.
+
+        The minimum frequency of the resulting CQT will be modified to
+        `fmin * 2**(tuning / bins_per_octave)`.
+
+    filter_scale : float > 0
```

Comment:
How does this play with the `gamma`?

Reply:
They're basically independent.  `filter_scale` is a multiplier on the Q factor, allowing for longer (or shorter) filters than would be dictated by the bandwidth calculation.  This was introduced very (very) long ago to match a feature of mcvicar's implementation.

Reply:
yes but isn't this going to lead us astray of ERB inadvertently?
if we tinker with the `filter_scale`, then this changes the bandwidth in inverse proportion, and thus changes the frequency where the ERB scales transitions from constant-bw to constant-Q. (because physiologically speaking, the ERB is an upper bound on the latency of the auditory system)
Ideally i'd like to see an equation in the documentation where all these elements are discussed in conjunction.

Reply:
Yeah, that's fair.  I'd actually be okay with deprecating this parameter and simplifying the whole thing.

Reply:
the way i like to think about it is the following: independently of the number of bins per octave and from the choice of wavelet, there are two degrees of freedom:
(1) `max_Q`: maximum quality factor. Defaults to `bins_per_octave`.
(2) `min_bandwidth`: minimum bandwidth, defaulting to zero.

If `max_Q < bin_per_octave`, there is an error.

The bandwidth of the filter at any frequency `f` is equal to:
(i) if `f < min_bandwidth`, then `f` (i.e. the quality factor is 1) 
(ii) otherwise, if `f` ranges between `min_bandwidth` and `max_Q*min_bandwidth`, then `min_bandwidth`
(iii) if `(f/max_Q) > min_bandwidth`, then `f/max_Q` (i.e. the quality factor is max_Q)

So there are actually three regimes: constant-Q, then constant-bw, then constant-Q again. The priorities are: Q>=1 first; bw > min_bw unless impossible; Q=max_Q unless impossible.

(in practice, we want a definition of bandwidth that is a little bit more whizzy than f/Q, involving 2^(1/Q)-2^(-1/Q) etc. but that doesn't change the narrative. also, we might want to smoothly interpolate between these regimes, rather than having sharp breakpoints)

just my 2c

Reply:
e.g. see here, where i expressed `min_bandwidth` in terms of its inverse (`max_scale`, measured in samples) for the convenience of the user.

https://github.com/lostanlen/WaveletScattering.jl/blob/master/src/spec.jl#L9

Reply:
here is where the construction happens
https://github.com/lostanlen/WaveletScattering.jl/blob/master/src/spec1d.jl#L55

Reply:
> involving 2^(1/Q)-2^(-1/Q)

Okay, let's resurface that conversation now, since we previously had it OTR.

Our current CQT filterbank uses `2**(1/bpo) - 1` to determine bandwidth, as given in the S&K 2010 paper.  However, the later paper that I'm basing this vqt implementation on uses `2**(1/bpo) - 2**(-1/bpo)`, effectively centering the frequency range rather than edge-aligning it.  I can see arguments for both, but should we expose an interface to control this?  Or is it way too far in the weeds for users to care about?

Reply:
For `bpo >> 1` and wavelets with a quasi-symmetric profile in the Fourier domain (i.e. almost every use case, including the default), it doesn't really make a difference. If we were making Gammatones with Q=2, then i would argue otherwise.

I don't see a compelling reason to break the current implementation